File: LeetCode/Study Plan/Practice Problems/Easy/Add To Array-Form Of Integer/Java/addToArrayFormOfInteger.java
class Solution {
    public List<Integer> addToArrayForm(int[] num, int k) {
        List<Integer> res = new ArrayList<>();
        int size = num.length;
        int i = size-1;
        int carry = 0;

        while (i >= 0 && k > 0) {
            int sum = num[i] + k%10 + carry;
            res.add(sum%10);
            carry = sum/10;
            i--;
            k/=10;
        }

        while(i >= 0) {
            int sum = num[i] + carry;
            res.add(sum%10);
            carry = sum/10;
            i--;
        }

        while(k > 0) {
            int sum = k%10 + carry;
            res.add(sum%10);
            carry = sum/10;
            k/=10;
        }
        
        while(carry > 0) {
            res.add(carry%10);
            carry/=10;
        }

        Collections.reverse(res);
        return res;
    }
}

File: LeetCode/Study Plan/Practice Problems/Easy/Add To Array-Form Of Integer/Go/addToArrayFormOfInteger.go
package main

import "fmt"

func addToArrayForm(num []int, k int) []int {
	size := len(num)
	res := []int{}
	i, carry := size-1, 0

	for i >= 0 && k > 0 {
		sum := num[i] + k%10 + carry
		res = append(res, sum%10)
		carry = sum / 10
		k /= 10
		i--
	}

	for i >= 0 {
		sum := num[i] + carry
		res = append(res, sum%10)
		carry = sum / 10
		i--
	}

	for k > 0 {
		sum := (k % 10) + carry
		res = append(res, sum%10)
		k = k / 10
		carry = sum / 10
	}

	for carry > 0 {
		res = append(res, carry%10)
		carry /= 10
	}

	return reverse(res)
}

func reverse(a []int) []int {
	size := len(a)
	for i := 0; i < size/2; i++ {
		a[i], a[size-1-i] = a[size-1-i], a[i]
	}
	return a
}

func main() {
	fmt.Println(addToArrayForm([]int{1, 2, 0, 0}, 34))
}
